src/metric/xentropy_metric.hpp

```diff
@@ -0,0 +1,384 @@
+ *
+ * See xentropy_objective.hpp for further details.
+ *
+ */
+
+namespace LightGBM {
+
+  // label should be in interval [0, 1];
+  // prob should be in interval (0, 1); prob is clipped if needed
+  inline static double XentLoss(float label, double prob) {
+    const double log_arg_epsilon = 1.0e-12;
+    double a = label;
+    if (prob > log_arg_epsilon) {
+      a *= std::log(prob);
+    } else {
+      a *= std::log(log_arg_epsilon);
+    }
+    double b = 1.0f - label;
+    if (1.0f - prob > log_arg_epsilon) {
+      b *= std::log(1.0f - prob);
+    } else {
+      b *= std::log(log_arg_epsilon);
+    }
+    return - (a + b);
+  }
+
+  // hhat >(=) 0 assumed; and weight > 0 required; but not checked here
+  inline static double XentLoss1(float label, float weight, double hhat) {
+    return XentLoss(label, 1.0f - std::exp(-weight * hhat));
+  }
+
+  // Computes the (negative) entropy for label p; p should be in interval [0, 1];
+  // This is used to presum the KL-divergence offset term (to be _added_ to the cross-entropy loss).
+  // NOTE: x*log(x) = 0 for x=0,1; so only add when in (0, 1); avoid log(0)*0
+  inline static double YentLoss(double p) {
+    double hp = 0.0;
+    if (p > 0) hp += p * std::log(p);
+    double q = 1.0f - p;
+    if (q > 0) hp += q * std::log(q);
+    return hp;
+  }
+
+//
+// CrossEntropyMetric : "xentropy" : (optional) weights are used linearly
+//
+class CrossEntropyMetric : public Metric {
+public:
+	explicit CrossEntropyMetric(const MetricConfig&) {}
+  virtual ~CrossEntropyMetric() {}
+
+  void Init(const Metadata& metadata, data_size_t num_data) override {
+    name_.emplace_back("xentropy");
+    num_data_ = num_data;
+    label_ = metadata.label();
+    weights_ = metadata.weights();
+
+    // ensure that labels are in interval [0, 1], interval ends included
+    if (label_ == nullptr) {
+       Log::Fatal("[%s]: label nullptr", __func__);
+    }
+    for (data_size_t i = 0; i < num_data; ++i) {
+    	if (label_[i] < 0.0f || label_[i] > 1.0f) {
+    		Log::Fatal("[%s]: metric does not tolerate label [#%i] outside [0, 1]", __func__, i);
+    	}
+    }
+    Log::Info("[%s:%s]: labels passed interval [0, 1] check (metric)",  GetName()[0].c_str(), __func__);
+
+    if (weights_ == nullptr) {
+      sum_weights_ = static_cast<double>(num_data_);
+    } else {
+      sum_weights_ = 0.0f;
+      for (data_size_t i = 0; i < num_data; ++i) {
+        sum_weights_ += weights_[i];
+      }
+    }
+    // check weight sum
+    if (sum_weights_ <= 0.0f) {
+      Log::Fatal("[%s]: sum-of-weights = %f is non-positive", __func__, sum_weights_);
+    } else {
+      Log::Info("[%s:%s]: sum-of-weights = %f",  GetName()[0].c_str(), __func__, sum_weights_);
+    }
+  }
+
+  std::vector<double> Eval(const double* score, const ObjectiveFunction* objective) const override {
+    double sum_loss = 0.0f;
+    if (objective == nullptr) {
+      Log::Warning("[%s : %d]: objective nullptr", __FILE__, __LINE__);

```

Comment:
What exactly is the situation where this is called without an objective function?
It seems there is not enough metadata (in general) to treat a score correctly for a metric unless you know how the objective function score is parameterized; easy to misapply metrics yes?